### PR TITLE
Move the semantics checks of Inheritance margin to OOP

### PR DESCRIPTION
**Pull request by Cosifne**
This PR is a cleanup PR and a potential solution for https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1528673/

Currently, the inheritance margin service would first get the declared symbols for the given document in devenv, then send all the symbols to OOP, then call FAR services and get the result back.
And as Cyrus mentioned in the issue, somehow this 'GetDeclarationSymbol' call is using a lot of memory. (maybe it hits some cache miss issue I guess)
Also, this is a little bit different from the 'Global imports' features (which is just sending a document and text span to the OOP), so it introduces some similar code.

This PR moves the 'GetDeclaredSymbol' part also to OOP and solves both problems.

### CODE REVIEW

File: src/Features/Core/Portable/InheritanceMargin/IRemoteInheritanceMarginService.cs [hunk 1]
```diff
@@ -12,18 +12,11 @@ namespace Microsoft.CodeAnalysis.InheritanceMargin
 {
     internal interface IRemoteInheritanceMarginService
     {
-        ValueTask<ImmutableArray<InheritanceMarginItem>> GetGlobalImportItemsAsync(
+        ValueTask<ImmutableArray<InheritanceMarginItem>> GetInheritanceMarginItemsAsync(
```
Comment (Cosifne) [inline src/Features/Core/Portable/InheritanceMargin/IRemoteInheritanceMarginService.cs:15]: This is the core part of the change.
For both features, just send a document and text span to the OOP, and do all the semantics work there.
So we only need one method in the remote interface.

File: src/Features/Core/Portable/InheritanceMargin/AbstractInheritanceMarginService_Helpers.cs [hunk 2]
```diff
@@ -150,6 +71,38 @@ internal abstract partial class AbstractInheritanceMarginService
...
+                    document: remapped ? null : document,
+                    symbolAndLineNumbers,
+                    frozenPartialSemantics: frozenPartialSemantics,
+                    cancellationToken).ConfigureAwait(false));
+            }
+
+            return result.ToImmutable();
+        }
+
         private async Task<ImmutableArray<InheritanceMarginItem>> GetGlobalImportsItemsInProcessAsync(
```
[the remark this answers is not in the document]
Comment (genlu) [reply src/Features/Core/Portable/InheritanceMargin/AbstractInheritanceMarginService_Helpers.cs:106]: Same for `GetSymbolInheritanceChainItemsInProcessAsync`

**Review by genlu — APPROVED** [review]
I love this change. One fewer feature that needs symbols in proc!

File: src/Features/Core/Portable/InheritanceMargin/IInheritanceMarginService.cs [hunk 3]
```diff
@@ -13,12 +13,15 @@ namespace Microsoft.CodeAnalysis.InheritanceMargin
...
-        /// Get the lines need to be have a margin and the member's information on that line.
+        /// Get information about
+        /// 1. The inheritance chain of the member in given span.
+        /// 2. The global imports for the document.
         /// </summary>
         ValueTask<ImmutableArray<InheritanceMarginItem>> GetInheritanceMemberItemsAsync(
             Document document,
             TextSpan spanToSearch,
             bool includeGlobalImports,
+            bool frozenPartialSemantics,
```
[the remark this answers is not in the document]
Comment (Cosifne) [reply src/Features/Core/Portable/InheritanceMargin/IInheritanceMarginService.cs:24]: This parameter is introduced in https://github.com/dotnet/roslyn/pull/61286
As Cyrus and Sam mentioned it is a strange pattern to indicate we are using the frozen partial semantics both in devenv and OOP.

>Does inheritance-margin need to operate on full semantic? 

See here https://sourceroslyn.io/#Microsoft.VisualStudio.LanguageServices/InheritanceMargin/InheritanceMarginTaggerProvider.cs,59
When the full semantic is ready the tagger would refresh all the tag again to make sure we get the latest semantics.
And we use the frozen partial one to make sure user could see it the glyph as early as possible

**Comment by CyrusNajmabadi** [comment]
so one thing the old system did was remap from a MAS workpsace to the normal workspaec (presumbly so inheritnace margin works in MAS files).  does that continue to work?

File: src/Features/Core/Portable/InheritanceMargin/IInheritanceMarginService.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (genlu) [reply src/Features/Core/Portable/InheritanceMargin/IInheritanceMarginService.cs:24]: I was wondering if the feature needs to operate **explicitly** on full-semantic. My understanding is, when full semantic is available, using frozen partial would do nothing special. The callsite in the tagger also just set `frozenPartialSemantics` to true, which was what prompted my question :)

**Comment by Cosifne** [comment]
> so one thing the old system did was remap from a MAS workpsace to the normal workspaec (presumbly so inheritnace margin works in MAS files). does that continue to work?

Are you asking this? https://sourceroslyn.io/#Microsoft.CodeAnalysis.Features/InheritanceMargin/AbstractInheritanceMarginService.cs,96

which is if a metadata symbol could be found in workspace, we should try to navigate to the 'in source' symbol?

There is a test covering that scenario, so I think it still works
https://sourceroslyn.io/#Microsoft.CodeAnalysis.EditorFeatures.UnitTests/InheritanceMargin/InheritanceMarginTests.cs,2109

File: src/Features/Core/Portable/InheritanceMargin/IInheritanceMarginService.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Cosifne) [reply src/Features/Core/Portable/InheritanceMargin/IInheritanceMarginService.cs:24]: >I was wondering if the feature needs to operate explicitly on full-semantic

No, we generate the InheritanceMarginItems whenever a partial semantic is ready.
And later when the full semantic is ready InheritanceMarginItems are generated again.

**Comment by CyrusNajmabadi** [comment]
Do your tests exercise the OOP paths as well as the non-oop paths?  Importantly, are tehy testing the MetadataAsSource scenario?

**Comment by Cosifne** [comment]
> Do your tests exercise the OOP paths as well as the non-oop paths? Importantly, are tehy testing the MetadataAsSource scenario?

I have changed all the tests to cover both in-process and OOP.

For the MetadataAsSource regression, as we talked offline, it is a problem probably related to DocumentIdAndSpan, not related to this PR.
I am going to check in this PR first and fix that problem later.

**Pull request merged**